pkg/resource/replication_group/post_set_output.go

```diff
@@ -88,3 +89,16 @@ func setMaintenanceWindow(
 		ko.Spec.PreferredMaintenanceWindow = &pmw
 	}
 }
+
+// setCacheParameterGroup updates the cache parameter group associated with the replication group
+//   (if non-nil in API response) regardless of whether it was specified in desired
+func setCacheParameterGroup(
+	latestCacheCluster *svcsdk.CacheCluster,
+	resource *resource,
+) {
+	ko := resource.ko
+	if latestCacheCluster.CacheParameterGroup != nil && latestCacheCluster.CacheParameterGroup.CacheParameterGroupName != nil {
+		cpgName := *latestCacheCluster.CacheParameterGroup.CacheParameterGroupName
+		ko.Spec.CacheParameterGroupName = &cpgName
```

Comment:
if `ko.Spec.CacheParameterGroupName` was non-nil and different from `latestCacheCluster.CacheParameterGroup.CacheParameterGroupName`, why would we overwrite `ko.Spec.CacheParameterGroupName` here? This smells funny...

Reply:
The intention is just to retrieve the cache parameter group currently associated to the replication group. That information comes from the `DescribeCacheClusters` API since it is not present in the `DescribeReplicationGroups` API response.

If `ko.Spec.CacheParameterGroupName` is non-nil and gets overwritten, that's precisely what we want because the original value was copied over from `desired`, the latest value is different from that, and the controller will go on to detect a diff and invoke a Modify call with the desired `CacheParameterGroupName`.

Reply:
https://github.com/aws-controllers-k8s/community/issues/846#issuecomment-868887930

This falls under "case 3" of spec fields as I mentioned in this issue a while back